keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,418 @@
+```
+
+Finally, when instance-specific configuration is also provided via the
+ComponentConfig approach, using the separate file proposed above.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: instance-config
+        image: something-with-bash
+        command:
+        - bash -c
+        args:
+        - cat << EOF > /etc/config/kube-proxy-instance.yaml
+          kind: KubeProxyInstanceConfiguration
+          apiVersion: kubeproxy.config.k8s.io/v1alpha1
+          hostnameOverride: ${NODE_NAME}
+          EOF
+        env:
+          - NODE_NAME
+            valueFrom:
+              fieldRef:
+                fieldPath: spec.nodeName
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --config=/etc/config/kube-proxy.yaml
+        - --instance-config=/etc/config/kube-proxy-instance.yaml
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Although the last example is more verbose, it provides the full benefits of
+the ComponentConfig approach, resulting in a consistent surface and versioning
+policy for the component's parameters.
+
+Sometimes, kube-proxy is configured as a static Pod instead of a DaemonSet.
+In that case, instance-specific configuration could be configured by the node
+startup scripts and mounted via a host path volume.
+
+### Kubelet Example
+
+The Kubelet is not typically deployed as a Pod. Instead, startup scripts that
+bootstrap the node write the Kubelet's initial configuration to disk at startup.
+In some deployments, the Kubelet's configuration may later be modified by the
+Dynamic Kubelet Config feature.
+
+Today, the startup script may write a Kubelet command line that looks something
+like `kubelet --config=/etc/config/kubelet.yaml --node-ip=${NODE_IP}`. Under
+this proposal, it would instead write both config files, and a command line like
+`kubelet --config=/etc/config/kubelet.yaml --instance-config=/etc/config/kubelet-instance.yaml`.
+
+When the Dynamic Kubelet Config feature is enabled, the Kubelet only replaces
+the values supplied to `--config` when consuming a remote configuration. Thus,
+the instance-specific parameters would not interfere with sharable parameters
+in this approach.
+
+### Test Plan
+
+ComponentConfig APIs are typically tested similarly to API server APIs, for
+example by testing round-trip conversions. Some tests specifically exercise
+the ability to share a config object, such as the Dynamic Kubelet Config tests.
+
+All existing tests for ComponentConfig Kinds should be extended to include the
+instance-specific Kinds, except where those tests are explicitly testing the
+ability to share a config object across concurrent instances.
```

Comment:
i think the test plan is fine as is. if a component wishes to op-in into adding instance specific config as per this KEP, it's their responsibility to e2e / integration test it.